controlplane/src/core/repositories/analytics/MetricsDashboardRepository.ts

```diff
@@ -0,0 +1,364 @@
+        params,
+      )}}[${prange}])) by (le))`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate,
+      }),
+    });
+
+    const prevResult = await this.client.queryRange({
+      query: `histogram_quantile(0.95, sum(rate(cosmo_router_http_request_duration_milliseconds_bucket{${this.getQueryLabels(
+        params,
+      )}}[${prange}] offset ${range}h)) by (le))`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate: endDate - range * 60 * 60 * 1000,
+      }),
+    });
+
+    return {
+      value: latencyP95.data.result[0]?.value[1],
+      previousValue: prevLatencyP95.data.result[0]?.value[1],
+      top: top5.data.result.map((v: any) => ({
+        name: v.metric.wg_operation_name,
+        value: v.value[1],
+      })),
+      series: this.mapSeries(endDate, range, result.data.result[0].values, prevResult.data.result[0]?.values),
+    };
+  }
+
+  /**
+   * Get error metrics
+   */
+  public async getErrorMetrics({ range = 24, endDate = getUtcUnixTime(), params }: GetMetricsProps) {
+    const prange = '15m';
+
+    // get error percentage of requests in last [range] hours
+    const percentage = await this.client.query({
+      query: `sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}, http_status_code!~"2.."}[${range}h])) / sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h])) * 100`,
+    });
+    const previousPercentage = await this.client.query({
+      query: `sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}, http_status_code!~"2.."}[${range}h])) / sum(rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h] offset ${range}h)) * 100`,
+    });
+
+    // get top 5 operations in last [range] hours
+    const top5 = await this.client.query({
+      query: `topk(5, sum by (wg_operation_name) (rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}, http_status_code!~"2.."}[${range}h])) / sum by (wg_operation_name) (rate(cosmo_router_http_requests_total{${this.getQueryLabels(
+        params,
+      )}}[${range}h])) * 100)`,
+    });
+
+    // get requests in last [range] hours in series of [step]
+    const result = await this.client.queryRange({
+      query: `sum (rate(cosmo_router_http_requests_total{${this.getQueryLabels(params)}}[${prange}]))`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate,
+      }),
+    });
+
+    const prevResult = await this.client.queryRange({
+      query: `sum (rate(cosmo_router_http_requests_total{${this.getQueryLabels(params)}}[${prange}]))`,
+      ...this.getRangeQueryProps({
+        range,
+        endDate: endDate - range * 60 * 60 * 1000,
+      }),
+    });
+
+    return {
+      value: percentage.data.result[0]?.value[1],
```

Comment:
That can throw. data might not be set when the request errors.

Reply:
Hmm then the types of the client are incorrect